f/rds external secret

chriskvariant opened the pull request:
# Description

Updated external secrets template to account for postgres-secrets that would be created by 'terraform-postgres-database'. Deployment.yaml for API also now mounts external secrets to a unique path in the case that more than one external secret is created.

Fixes [#8](https://usxtech.atlassian.net/browse/CLOUD-8)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

This was tested by utilizing the 'coolapi' helm install commands to create an application. The external secret resources were monitored and tested using this method to determine if I was getting the expected output.

The Octopus deployment for integration testing also succeeded: https://octopus.apps.ops-drivevariant.com/app#/Spaces-2/projects/lazy-helm-charts/deployments/releases/0.1.0-f-rds-external-s0001.490/deployments/Deployments-41268?activeTab=taskSummary

Pre reqs:
  - Check kube context and ensure you are in DPL or run through terminal in Lens in the DPL cluster
  - Use the below values.yaml. It is identical to the 'default-values.yaml' in lazy-helm-charts api CI other than the addition of two more secrets.
```
deployment:
  image:
    tag: 064859874041.dkr.ecr.us-east-2.amazonaws.com/lazy-helm-charts/variant-api:v0.3
revision: testing

istio:
  enabled: true
  ingress:
    host: dpl-drivevariant.com
service:
  targetPort: 5000

configVars:
  PLAIN_CONFIG: foo

secretVars:
  SECRET_CONFIG: secret
  SOMETHING: else

configMaps:
  - test-existing

awsSecrets:
  - name: devops-playground-rds-creds
  - name: postgres-secret-test-db-36
  - name: postgres-secret-test-db-inverse-05


livenessProbe:
  httpGet:
    path: /health
    port: 5000

readinessProbe:
  httpGet:
    path: /health
    port: 5000

tags:
  octopus/project: test
  octopus/pace: default
  octopus/environment: dev
  octopus/project_group: default group
  octopus/release_channel: test1
  team: devops
  purpose: test
  owner: devops
``` 

Test Case 1 [Verifying Processing of Postgres and Non-Postgres Secrets]:
- Steps
  - `helm repo add variant-inc-helm-charts https://variant-inc.github.io/lazy-helm-charts/ --force-update`
  - `helm upgrade --install sre-test-rds-secrets -n demo variant-inc-helm-charts/variant-api --version 2.1.13-beta -f values.yaml --dry-run`
  - `helm upgrade --install sre-test-rds-secrets -n demo variant-inc-helm-charts/variant-api --version 2.1.13-beta -f values.yaml`
- Result
    - A new pod should be created using the values.yaml and the 2.1.15 beta version of lazy-helm-charts.
    - This one test case should cover essentially all aspects of my change, as it tests: 
        - How postgres specific secrets are handled
        - How non-postgres secrets are processed as they were before
        - That multiple secrets can be added and mounted to the pod using volume mounts
        - That postgres AWS secrets are now added to the pod as environment variables as a unique key value pair
 
After creating the beta release, the beta release(2.1.15-beta) was run manually in DPL using the previous commands.

Test Case 2 [Creating DX dummy applications using beta release for every chart]
- Steps
  - [CI Build] https://github.com/variant-inc/demo-python-flask-variant-api/actions/runs/2701352019
- Results
  - The following applications were created with the values.yamls listed in a test branch in demo-python-flask-variant-app(https://github.com/variant-inc/demo-python-flask-variant-api/tree/test/rds-secrets-chris/.variant/deploy)
  - sre-test-api, sre-test-cron, sre-test-ui, and sre-test-handler should all reflect the below changes
    - All of the applications should have the pod volume mounts as in the image below(secret names may differ)
![image](https://user-images.githubusercontent.com/97901744/179887659-e5d9bfc9-29e4-4563-9c40-602745eae3bd.png)

    - All of the pods should have postgres secrets set as environment variables in the pod and have their names formatted specifically as seen in the image below
![image](https://user-images.githubusercontent.com/97901744/179888344-8440cb46-7161-4994-bf53-c037de568184.png)

    - All of the applications can be found in the DPL cluster in the SRE name space. Furthermore, here is the link to all of the octopus deployments
   
      - https://octopus.apps.ops-drivevariant.com/app#/Spaces-2/projects/sre-test-api
      - https://octopus.apps.ops-drivevariant.com/app#/Spaces-2/projects/sre-test-cron
      - https://octopus.apps.ops-drivevariant.com/app#/Spaces-2/projects/sre-test-handler
      - https://octopus.apps.ops-drivevariant.com/app#/Spaces-2/projects/sre-test-ui 

- [x] Sanity Testing

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

genghis-tuan commented on charts/library/templates/_external-secret.tpl:
```diff
@@ -2,17 +2,31 @@
+      {{.name}}-host: "<%= JSON.parse(data.data).host %>"
+      {{.name}}-dbname: "<%= JSON.parse(data.data).dbname %>"
+      {{.name}}-username: "<%= JSON.parse(data.data).username %>"
+      {{.name}}-passowrd: "<%= JSON.parse(data.data).password %>"
```
is 'passowrd' intentional?

genghis-tuan replied:
Sorry, I meant the spelling.

chriskvariant replied:
Ah, gotcha - good catch! The mis-spelling should not have affected any functionality, as the misspelled word was the output variable name. I have fixed the spelling.

chriskvariant commented:
@genghis-tuan I also have not updated the version out of beta; let me know when I should go ahead and make that change.

genghis-tuan reviewed (commented):
Hey Chris, would you mind providing your testing files such as the values.yaml in the testing steps?  Also, it would be easier for testing and reviewing if you formatted the testing steps like the example below so that it is clear to the review how to replicate your results.

```yaml
Pre reqs:
  - any prerequisites such as files, repos, settings, etc
Test Case 1 [What is this test case verifying]:
- Steps
  - executed this command
  - modified this file
  - did something else
- Result
 - something was changed, created, deleted, etc...
 - something else

Test Case 2 [Verifying something else]:
- Steps
  - did this
- Results
  - something happened
... 
```

chriskvariant commented:
> Hey Chris, would you mind providing your testing files such as the values.yaml in the testing steps? Also, it would be easier for testing and reviewing if you formatted the testing steps like the example below so that it is clear to the review how to replicate your results.
> 
> ```yaml
> Pre reqs:
>   - any prerequisites such as files, repos, settings, etc
> Test Case 1 [What is this test case verifying]:
> - Steps
>   - executed this command
>   - modified this file
>   - did something else
> - Result
>  - something was changed, created, deleted, etc...
>  - something else
> 
> Test Case 2 [Verifying something else]:
> - Steps
>   - did this
> - Results
>   - something happened
> ... 
> ```

Thanks for the suggestion Tuan! I updated the original PR post in the format you gave an example of. Please let me know if you would like me to add more test cases and I will be happy to do so!

salyateem13 commented:
@chriskvariant I'm trying to test, but the external secret controller isn't find the secret in secret manager. do I need to create the secret

salyateem13 commented:
<img width="674" alt="Screen Shot 2022-07-14 at 9 30 56 AM" src="https://user-images.githubusercontent.com/31292258/178994281-8777f6b9-2579-4e31-9855-5eb80bcead77.png">

Actually, I ran it in the wrong cluster by mistake. I ran the testing steps, and am seeing the secret get created, but are the template values supposed to be interpolated at the helm step? Or when they get mounted to the pod? But I can't see the secret values attached to the cool api pod. Not sure what I'm missing

salyateem13 commented on charts/library/templates/_external-secret.tpl:
```diff
@@ -2,17 +2,31 @@
+      {{.name}}-dbname: "<%= JSON.parse(data.data).dbname %>"
+      {{.name}}-username: "<%= JSON.parse(data.data).username %>"
+      {{.name}}-password: "<%= JSON.parse(data.data).password %>"
+      {{.name}}-engine: "<%= JSON.parse(data.data).engine %>"
```
Can we have the env variable keys match those of the configmap its going to replace.

chriskvariant replied:
Absolutely - I have made the change [here](https://github.com/variant-inc/lazy-helm-charts/blob/d42051b2bf8c0d9f379f50808694ab27317a370e/charts/library/templates/_external-secret.tpl#L21).
Please let me know if this is correct!

chriskvariant commented:
The original PR information has been updated with more information regarding the testing process.

genghis-tuan reviewed (approved):
Looks great, Chris. I see the external secrets in the pod details for all 4 charts and also see them in the environment variables. Can you please locate the [lazy-helm release for this build](https://octopus.apps.ops-drivevariant.com/app#/Spaces-2/projects/lazy-helm-charts/deployments/releases/0.1.0-f-rds-external-s0001.508) and deploy it to satisfy the PR requirements. Then fix any conflicts, squash, bump the versions to remove '-beta' and merge.
